Source/CarthageKit/VersionFile.swift

```diff
@@ -216,47 +237,42 @@ public func createVersionFileForCommitish(
+
+	var platformCaches: PlatformCaches = [:]
 
 	let platformsToCache = platforms.isEmpty ? Set(Platform.supportedPlatforms) : platforms
 	for platform in platformsToCache {
 		platformCaches[platform.rawValue] = []
 	}
 
-	let writeVersionFile = SignalProducer<(), CarthageError> { () -> Result<(), CarthageError> in
-		let rootBinariesURL = rootDirectoryURL.appendingPathComponent(Constants.binariesFolderPath, isDirectory: true).resolvingSymlinksInPath()
-		let versionFileURL = rootBinariesURL.appendingPathComponent(".\(dependencyName).\(VersionFile.pathExtension)")
-
-		let versionFile = VersionFile(
-			commitish: commitish,
-			macOS: platformCaches[Platform.macOS.rawValue],
-			iOS: platformCaches[Platform.iOS.rawValue],
-			watchOS: platformCaches[Platform.watchOS.rawValue],
-			tvOS: platformCaches[Platform.tvOS.rawValue])
-
-		return versionFile.write(to: versionFileURL)
-	}
-
 	if !buildProducts.isEmpty {
 		return SignalProducer<URL, CarthageError>(buildProducts)
-			.flatMap(.merge) { url -> SignalProducer<String, CarthageError> in
-				let platformName = url.deletingLastPathComponent().lastPathComponent
+			.flatMap(.merge) { url -> SignalProducer<(String, (String, String)), CarthageError> in
 				let frameworkName = url.deletingPathExtension().lastPathComponent
+				let platformName = url.deletingLastPathComponent().lastPathComponent
 				let frameworkURL = url.appendingPathComponent(frameworkName, isDirectory: false)
-				return hashForFileAtURL(frameworkURL)
-					.on(value: { hash in
-						let cachedFramework = CachedFramework(name: frameworkName, hash: hash)
-						if var frameworks = platformCaches[platformName] {
-							frameworks.append(cachedFramework)
-							platformCaches[platformName] = frameworks
-						}
-					})
+				let details = SignalProducer<(String, String), CarthageError>(value: (platformName, frameworkName))
+				return SignalProducer.zip(hashForFileAtURL(frameworkURL), details)
+			}
+			.reduce(into: platformCaches) { (platformCaches: inout PlatformCaches, values: (String, (String, String))) in
+				let hash = values.0
+				let platformName = values.1.0
+				let frameworkName = values.1.1
+
+				let cachedFramework = CachedFramework(name: frameworkName, hash: hash)
+				if var frameworks = platformCaches[platformName] {
+					frameworks.append(cachedFramework)
+					platformCaches[platformName] = frameworks
+				}
+			}
+			.flatMap(.merge) { platformCaches -> SignalProducer<(), CarthageError> in
+				createVersionFile(commitish, dependencyName: dependencyName, rootDirectoryURL: rootDirectoryURL, platformCaches: platformCaches)
 			}
-			.then(writeVersionFile)
 	} else {
 		// Write out an empty version file for dependencies with no built frameworks, so cache builds can differentiate between
 		// no cache and a dependency that has no frameworks
-		return writeVersionFile
+		return createVersionFile(commitish, dependencyName: dependencyName, rootDirectoryURL: rootDirectoryURL, platformCaches: platformCaches)
```

Comment:
Can we pass `[:]` instead of `platformCaches` here to make it obvious that it's an empty list?

Then we can move the declaration of `platformCaches` to be closer to where it's used.

Reply:
Good point!